Taint node with NoSchedule effect when ovnkube pod is down

tssurya opened the pull request:
- What this PR does and why is it needed

UseCase: During upgrades, when the ovnkube pod is deleted and
before/while it gets recreated if a new pod request comes, it
will fail.

When node networking is down, we should not allow pods to be created
on that node. This PR fixes this by adding a taint on the node with
the NoScedule effect when the ovnkube pod receives a SIGTERM and then
removes the taint when the ovnkube pod finishes setting up networking
and initializing the CNI config.

- How to verify it
The added taint will be visible on the node for the brief period of time between the deletion of ovnkube-node pod and creation of the new one that takes its place which roughly takes ~60sec.

```
I0718 09:41:21.603487  396533 kube.go:161] Added taint k8s.ovn.org/network-unavailable:NoSchedule on node ci-ln-sz84n3b-f76d1-h69qs-worker-b-q2vsm
Taints:             k8s.ovn.org/network-unavailable:NoSchedule
I0718 09:12:24.775588  396533 kube.go:191] Removed taint k8s.ovn.org/network-unavailable:NoSchedule on node ci-ln-sz84n3b-f76d1-h69qs-worker-b-q2vsm
Taints:             <none>
```

- Description for the changelog
Taint node with NoSchedule effect when ovnkube pod is down and remove the taint when the pod is available again

tssurya commented on go-controller/cmd/ovnkube/ovnkube.go:
```diff
@@ -121,6 +133,26 @@ func main() {
 		select {
 		case s := <-exitCh:
 			klog.Infof("Received signal %s. Shutting down", s)
+			if s == syscall.SIGTERM {
```
note: There are a lot of nested ifs that I am not a fan of. However since we shouldn't really care to error out if we fail in one of the steps, I did this. But if folks feel I need to split them out to make it eye-friendly I can do that as well.

dcbw replied:
Yeah, I'd pull this out into a separate function, and perhaps collapse the ifs by doing stuff like:

```
if ovnkclient, err := util.NewOVNClientset(&config.Kubernetes); err != nil {
    return fmt.Errorf(xxx)
}

if nodeObj, err := ovnkclient.KubeClient.CoreV1().Nodes().Get(context.TODO(), node, metav1.GetOptions{}); err != nil {
    return fmt.Errorf(xxxx)
}

etc.
```

But this probably should also go into pkg/node/node.go as a goroutine that Start() calls. It can add 1 to the WaitGroup passed to Start() to make sure the node process doesn't exit until we've added the taint. Then it can "<-n.stopChan" and after that do all the taint addition. The benefit of that is you already have a KubeClient near the top of Start() that you cna use:

```
	if node, err = n.Kube.GetNode(n.name); err != nil {
		return fmt.Errorf("error retrieving node %s: %v", n.name, err)
	}
```

tssurya replied:
makes sense. Let me do that.

tssurya replied:
done.

dcbw commented on go-controller/pkg/node/node.go:
```diff
@@ -468,6 +473,19 @@ func (n *OvnNode) Start(wg *sync.WaitGroup) error {
 	}
 
+	// remove the NoSchedule Taint from the node, now that networking setup is done.
+	if nodeObjJson, err := json.Marshal(node); err == nil {
```
@tssurya I think I'd like to see these (add taint, remove taint) as separate functions with the `if x, err := Foo(); err != nil { return }` pattern. Makes the already long code in Start() more readable.

tssurya replied:
done.

dcbw commented:
Lastly, could you do the vendor updates as a separate commit? That  makes it easier to review the actual logic. Thanks!

tssurya commented:
> Lastly, could you do the vendor updates as a separate commit? That makes it easier to review the actual logic. Thanks!

yup I can do the vendoring as a separate 1st commit in the same PR/

dcbw commented:
@tssurya any updates on the vendor as separate commit? Thanks!

astoycos commented:
/assign @astoycos

squeed commented on go-controller/cmd/ovnkube/ovnkube.go:
```diff
@@ -14,7 +15,12 @@ import (
+	apimachinerytypes "k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/strategicpatch"
 	"k8s.io/klog/v2"
+	"k8s.io/kubernetes/pkg/util/taints"
```
We shouldn't vendor in `k8s.io/kubernetes` - can this be copy-pasted?

tssurya replied:
I don't follow, we already have the `k8s.io/kubernetes` vendored.. I am just trying to add this package taints similar to what we do in SDN.

tssurya commented:
> @tssurya any updates on the vendor as separate commit? Thanks!

sorry missed this comment. The vendoring has already been split into a separate commit (locally). I am doing some final testing for the actual code based on your comments. Will push an update at my CET TZ tomorrow. Extremely sorry for the delay, got stuck with other work but this is progressing well.

squeed commented on go-controller/go.mod:
```diff
@@ -40,6 +40,7 @@ require (
 	k8s.io/client-go v0.20.0
 	k8s.io/klog v0.0.0-20181102134211-b9b56d5dfc92
 	k8s.io/klog/v2 v2.4.0
+	k8s.io/kubernetes v0.0.0-00010101000000-000000000000
```
hey @tssurya this shows that we're vendoring k/k for the first time here...

tssurya replied:
oh heh you are right, https://github.com/ovn-org/ovn-kubernetes/tree/master/go-controller/vendor/k8s.io sorry. So I shouldn't do it this way?

astoycos replied:
@tssurya I think what Casey was trying to say is that it's not good practice to vendor K8s/k8s, and basically we shouldn't do it..... Instead I think we should just copy the k8s.io/kubernetes/pkg/util/taints package into our codebase manually.  I.e Add a new folder in /pkg/util and put the taints code there

astoycos replied:
(But don't quote me on it :) )

tssurya replied:
yea we discussed this in the meeting. Considering I need to copy a lot of dependent functions, decision as to move forward with vendoring.

danwinship replied:
You don't _have to_ use `pkg/util/taints`. Taints are part of the public API of v1.Node; as long as you are setting/clearing them correctly, you can do it "by hand". The utility functions are just there to make things easier, but if you end up having to vendor in all of k/k, that's not really "easier" any more.

tssurya commented:
PR is tested and checked. Before ovnkube pod receives SIGKILL we add the taint:
`Taints:             k8s.ovn.org/network-unavailable=NoSchedule:NoSchedule`

I0505 08:48:47.434398  711221 node.go:556] Successful in adding k8s.ovn.org/network-unavailable taint on node ci-ln-bhn0z9b-f76d1-chfdz-master-1

After a new pod takes its place we remove the taint on the node:
Taints:             <none>

For master nodes, in addition to the normal noschedule taint, we will have the unavailable-network taint as well.

tssurya commented:
Unit test failure seems unrelated. Checking it out.
```
/home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/ovn/policy_test.go:1235
Timed out after 1.001s.
Expected
    <bool>: false
to be true
/home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/ovn/address_set/fake_address_set.go:151
```

tssurya commented:
/retest

tssurya commented:
@dcbw : I've addressed your comments. PTAL. Sorry for the delay in getting to this.

The only confusion now is regarding the vendoring. Casey mentioned I should probably copy all the taint related functions I need instead of vendoring. Not sure if I should do that because then I'll need to copy a lot of dependent functions from that package. I'd guess we'd need more `k/k` packages in future so vendoring it in now should be ok but I could be wrong.

astoycos commented on go-controller/pkg/node/node.go:
```diff
@@ -478,6 +501,62 @@ func (n *OvnNode) Start(wg *sync.WaitGroup) error {
+	node, err := n.Kube.GetNode(n.name)
+	networkUnavailableTaint := &kapi.Taint{Key: "k8s.ovn.org/network-unavailable", Value: "NoSchedule", Effect: "NoSchedule"}
+
+	if err != nil {
```
nit: put error checking directly after `GetNode` and before assigning the `networkUnavailableTaint`

astoycos commented:
generally looks good @tssurya I think vendoring K/K is just bad practice, In general I've seen that it usually breaks some facet of `go mod` due to its size/dependency loops. Why not just copy the taint package out to our `pkg/util` package?

danwinship commented on go-controller/pkg/node/node.go:
```diff
@@ -478,6 +501,62 @@ func (n *OvnNode) Start(wg *sync.WaitGroup) error {
+	}
+
+	node, err := n.Kube.GetNode(n.name)
+	networkUnavailableTaint := &kapi.Taint{Key: "k8s.ovn.org/network-unavailable", Value: "NoSchedule", Effect: "NoSchedule"}
```
use the upstream taint: `corev1.TaintNodeNetworkUnavailable`

tssurya replied:
I can't really do this, for some reason something else is removing this taint as soon as ovnkube-node adds this. I don't know what is removing this.

creydr commented on go-controller/pkg/node/node.go:
```diff
@@ -478,6 +501,62 @@ func (n *OvnNode) Start(wg *sync.WaitGroup) error {
+		return err
+	}
+
+	nodeObjJson, err := json.Marshal(node)
+	if err != nil {
+		klog.Infof("Unable to marshal original node %s object for %s taint: %v", n.name, action, err)
+		return err
+	}
+
+	var taintedNodeObj *kapi.Node
+	if setTaint {
+		taintedNodeObj, _, err = taints.AddOrUpdateTaint(node, networkUnavailableTaint)
+	} else {
+		taintedNodeObj, _, err = taints.RemoveTaint(node, networkUnavailableTaint)
+	}
+
+	if err != nil {
+		klog.Infof("Unable to (un)taint node %s: %v", n.name, err)
+		return err
+	}
+
+	taintedNodeObjJson, err := json.Marshal(taintedNodeObj)
+	if err != nil {
+		klog.Infof("Unable to marshal updated node %s object for %s taint: %v", n.name, action, err)
+		return err
+	}
+
+	patchBytes, err := strategicpatch.CreateTwoWayMergePatch(nodeObjJson, taintedNodeObjJson, kapi.Node{})
+	if err != nil {
+		klog.Infof("Unable to patch the updated node %s object for %s taint: %v", n.name, action, err)
+		return err
+	}
+
+	if _, err = n.client.CoreV1().Nodes().Patch(context.TODO(), n.name, apimachinerytypes.StrategicMergePatchType, patchBytes, metav1.PatchOptions{}); err != nil {
+		klog.Infof("Unable to patch the updated node %s object for %s taint: %v", n.name, action, err)
+		return err
+	}
```
in #2326 I am introducing https://github.com/ovn-org/ovn-kubernetes/blob/45f51f0cc8ee83c19f61bec89d568588055f0ee0/go-controller/pkg/kube/kube.go#L223 and https://github.com/ovn-org/ovn-kubernetes/blob/45f51f0cc8ee83c19f61bec89d568588055f0ee0/go-controller/pkg/kube/kube.go#L234 which could be used for this.

tssurya replied:
ty! I am copying the change and will send out the PR shortly!

creydr replied:
Hey @tssurya,
you don't need to copy the change. You can simply cherry-pick the commit. Otherwise I can simply refactor your code after I got your changes for the taint packages.

tssurya replied:
@creydr : no that's not what I meant :) I have already copied the code from k8s taints package and added it to ovn-k. We can both use those functions and now do not need to vendor k8s.

tssurya replied:
I am testing this on OCP, once it passes will push it here.

tssurya commented:
@squeed @dcbw PTAL when you have time.

tssurya commented on go-controller/pkg/kube/kube.go:
```diff
@@ -23,6 +25,9 @@ type Interface interface {
 	SetAnnotationsOnPod(namespace, podName string, annotations map[string]string) error
 	SetAnnotationsOnNode(node *kapi.Node, annotations map[string]interface{}) error
 	SetAnnotationsOnNamespace(namespace *kapi.Namespace, annotations map[string]string) error
+	SetTaintOnNode(nodeName string, taint *kapi.Taint) error
```
Note: These bits (set, remove patch taints) are also needed for the work being done in https://github.com/ovn-org/ovn-kubernetes/pull/2326. I have tried to align it as much as possible to what is required there so that both can leverage it, but a few things might need to be changed for the other PR or vice versa based on which gets in first.

dcbw commented:
/lgtm
